### ISSUE TITLE
The behavior for omitted runAsUser is not clear and insecure.

### ISSUE THREAD

**Issue by erkanerol** [kubernetes/website#14574]
**This is a Bug Report**

**Problem:**

***Short Version***
With k8s 1.13.6, when I don't set runAsUser in securityContext, the container is started with root(0) user even when the default user of the container is different. This issue does not exist in 1.13 doc. Also, I believe this behavior is insecure. 

***Long Version***
When I am trying to test my app with k8s 1.13.6, I encountered some issues regarding runAsUser field. I checked the documentation and saw this part. 

https://github.com/kubernetes/website/blame/master/content/en/docs/tasks/configure-pod-container/security-context.md#L54

When I checked 1.13 doc, I didn't see this part. See: https://v1-13.docs.kubernetes.io/docs/tasks/configure-pod-container/security-context/

However, as I said, I am using k8s1.13.6. 

Now, the problems:
- There is no direct mention for the omitted runAsUser field. 
- If it is set as root(0) when it is omitted for k8s 1.13 too, k8s 1.13 doc should contain this behavior.
- Are you really sure about this behavior? I believe default behavior should be more restrictive. Image owners may set a different user as the default user for some security concerns and k8s -as container orchestrator- should not change the behavior in a lower layer.

**Proposed Solution:**
This behavior should be clarified in the k8s docs. Also, the behavior should be reconsidered again.

**Page to Update:**
https://kubernetes.io/docs/tasks/configure-pod-container/security-context/

***How to reproduce this issue***

I built a new docker image for testing this issue. I set the default user of `alpine` image as postgres.
Dockerfile: https://hub.docker.com/r/erkanerol/alpine/dockerfile

The default behaviors of the images
```
# docker run -it  alpine
/ # whoami
root

# docker run -it  erkanerol/alpine:postgres
/ $ whoami
postgres
```
I created two pod: one for `alpine`, one for `erkanerol/alpine:postgres`

postgres-pod.yaml
```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    run: whoami-alpine-postgres
  name: whoami-alpine-postgres
spec:
  containers:
  - command:
    - whoami
    image: erkanerol/alpine:postgres
    imagePullPolicy: IfNotPresent
    name: whoami-alpine-postgres
    resources: {}
  dnsPolicy: ClusterFirst
  restartPolicy: Never
status: {}
```

root-pod.yaml
```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    run: whoami-alpine-root
  name: whoami-alpine-root
spec:
  containers:
  - command:
    - whoami
    image: alpine
    imagePullPolicy: IfNotPresent
    name: whoami-alpine-root
    resources: {}
  dnsPolicy: ClusterFirst
  restartPolicy: Never
status: {}
```

Test 1 with k8s1.13
```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.3", GitCommit:"a4529464e4629c21224b3d52edfe0ea91b072862", GitTreeState:"clean", BuildDate:"2018-09-09T18:02:47Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.6", GitCommit:"abdda3f9fefa29172298a2e42f5102e777a8ec25", GitTreeState:"clean", BuildDate:"2019-05-08T13:46:28Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}

# kubectl apply -f postgres-pod.yaml
pod/whoami-alpine-postgres created

# kubectl logs pod/whoami-alpine-postgres
root   ->> (IT SHOULD BE postgres)

# kubectl apply -f root-pod.yaml
pod/whoami-alpine-root created

# kubectl logs pod/whoami-alpine-root
root
```

Test2 with k8s1.10
```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.3", GitCommit:"a4529464e4629c21224b3d52edfe0ea91b072862", GitTreeState:"clean", BuildDate:"2018-09-09T18:02:47Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.9", GitCommit:"e6ab4ee54f71e7a403cc733534bab86fd959ecb7", GitTreeState:"clean", BuildDate:"2018-10-16T12:01:43Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

# kubectl apply -f postgres-pod.yaml
pod/whoami-alpine-postgres created

# kubectl logs pod/whoami-alpine-postgres
postgres

# kubectl apply -f root-pod.yaml
pod/whoami-alpine-root created

# kubectl logs pod/whoami-alpine-root
root
```

containerRuntimeVersion is docker 18.6.1 for both tests.

**Comment by sftim** [comment]
@erkanerol - when you've written `v1.3.6` or `1.3.6`, did you actually mean `1.13.6`?

Can you confirm what container runtime you're using on nodes when you're testing this (eg, Docker)?

**Comment by erkanerol** [comment]
Hi. Sorry for typo. Yes it is 1.13.6 and containerRuntimeVersion is docker 18.6.1. I updated the description.

**Comment by sftim** [comment]
It sounds like the 1.13 cluster isn't honoring the `User` [property](https://github.com/opencontainers/image-spec/blob/master/config.md#properties) from the OCI image spec, whereas the 1.10 cluster does honor it. 

I didn't see anything mentioned in changelogs around this. Seems like something that really would be a surprise to a cluster operator.

**Comment by sftim** [comment]
_(Without looking much into this, and definitely without bringing up clusters to test it)_
I'm wondering if Pod plus `SecurityContext` plus the defaults for `runAsUser` have led to this behaviour.

If `runAsUser` is unset, should than mean UID 0? Or should it mean to start a process as the image's default user (as set via image properties).

If so, I suppose the same would go for `runAsGroup`, because the `config.User` property can specify a default group.

**Comment by erkanerol** [comment]
@sftim Do you need anything else from my side?

**Comment by sftim** [comment]
@erkanerol from my point of view this issue teport makes sense and explains the situation. The steps from this point, I'm not sure about.

**Comment by sftim** [comment]
This is due to a Kubernetes bug; see https://github.com/kubernetes/kubernetes/issues/78308 and https://nvd.nist.gov/vuln/detail/CVE-2017-11245

**Comment by liggitt** [comment]
The behavior in v1.14.2 and v1.13.6 is not intended. Omitting runAsUser *should* run the container using the USER specified in the container image.

**Comment by sftim** [comment]
Kubernetes v1.13.7 and v1.14.3 have fixes for this bug.
Now that's done: @erkanerol you can get the intended, documented behavior by upgrading to those releases.

/triage support

**Comment by erkanerol** [comment]
Hi. Thanks for the quick fix! I have not a cluster with 1.13.7 now. Will check soon and give an update.

**Comment by sftim** [comment]
Documentation is correct (and Kubernetes bug is fixed)
/close
